proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -293,70 +262,234 @@ protocol TargetBuildContext {
+        let name: String
         
         /// Path of the built or provided tool in the file system.
-        var path: Path { get }
+        let path: Path
+    }
+}
+```
+
+##### Package Model
+
+The context's `package` property references a directed acyclic graph of `Package`, `Product`, and `Target` objects.
+
+```swift
+/// Represents a package in the graph (either the root or a dependency).
+class Package {
+    /// The name of the package (for display purposes only).
+    let name: String
+
+    /// The absolute path of the package directory in the local file system.
+    let directory: Path
+
+    /// Any dependencies on other packages, in the same order as they are
+    /// specified in the package manifest.
+    let dependencies: [Dependency]
+
+    /// Represents a resolved dependency of a package on another package.
+    /// Although there is only one property today, future versions of the
+    /// API may add more information about the dependency here.
+    struct Dependency {
+        /// The package to which the dependency was resolved.
+        let package: Package
+    }
+    
+    /// Any regular products defined in this package (i.e. excluding plugin
+    /// products), in the same order as specified in the package manifest.
+    let products: [Product]
+
+    /// Any regular targets defined in this package (i.e. excluding plugin
+    /// targets), in the same order as specified in the package manifest.
+    let targets: [Target]
+}
+
+/// Represents a product defined in a package. Specializations represent
+/// different types of products.
+class Product {
+    /// The name of the product, as defined in the package manifest. It is
+    /// unique among the products of the package in which it is defined.
+    let name: String
+    
+	  /// The targets that directly comprise the product, in the order in
+    /// which they are declared in the package manifest.
+    let targets: [Target]
+}
+
+/// Represents an executable product defined in a package.
+class ExecutableProduct: Product {
+    /// The target that contains the main entry point of the executable.
+    /// Every executable product has exactly one main executable target.
+    /// This target will always be one of the targets in the product's
+    /// `targets` array.
+    let mainTarget: Target
+}
+
+/// Represents a library product defined in a package.
+class LibraryProduct: Product {
+    /// Whether the library is static, dynamic, or whether the type is
+    /// determined automatically.
+    let type: LibraryType
+
+    /// A type of library product.
+    enum LibraryType {
+        /// Static library.
+        case `static`
+
+        /// Dynamic library.
+        case `dynamic`
+
+        /// The type of library is unspecified and will be determined
+        /// at build time.
+        case automatic
+    }
+}
+
+/// Represents a target defined in a package. Specializations represent
+/// different types of targets.
+class Target {
```

Comment:
Do we need to represent supported platforms here (from conditional target dependencies)?

Reply:
Not in this initial release, I think — I have in fact pared this back a bit from where I had taken it so it's a smaller amendment that just focused on the `@main` without adding a bunch of extras.